src/tree/updater_quantile_hist.cc

```diff
@@ -706,6 +729,8 @@ void QuantileHistMaker::Builder<GradientSumT>::InitSampling(const std::vector<Gr
   unused_rows_.resize(info.num_row_);
   size_t* p_row_indices_used = row_indices->data();
   size_t* p_row_indices_unused = unused_rows_.data();
+  std::vector<GradientPair>& gpair_ref = const_cast<std::vector<GradientPair>&>(gpair);
```

Comment:
Can you just make a copy of gradient as a class member and use it through out the current iteration?

Reply:
copy is expensive operation for each iteration, and hopefully there is no  need to do a copy for single tree case (when `num_parallel_tree == 1`).
So currently I simplified the `InitSampling` (no dependency on number of trees) and copy is done only for not single trees (`num_parallel_tree != 1`).